handout/architect/10_ways_to_misuse_angular.md

```diff
@@ -0,0 +1,439 @@
+
+```ts
+@Component({
+  selector: 'my-foo',
+  host: { '[class]': 'className' },
+})
+export class FooComponent {
+  className = 'navy';
+}
+```
+
+
+
+## 2. Avoiding observables
+When working with Angular you should be using observables as much as possible. Angular exposes HTTP requests, form events, route params, etc. as observables. Therefore, by using observables we can build apps that have a consistent architectural pattern.
+
+### Leverage Observable Operators
+We can leverage observable operators to compute data based on application state. This allows us to reduce the amount of state we need to maintain. Additionally these computations can expressed as pure functions which makes testing easier. These pure play a similar to role to [selectors](https://github.com/reactjs/reselect) and can be memoized.
+
+This pattern can be adopted regardless of whether you are using ngrx, angular-redux or even just services to manage state. When woking with Angular services you can replicate a _store_ style API using Behavior Subjects. In the example below we have a `TodoStore` service. It exposes the entire store as the `state$` observable. To display a list of filtered todos we can simply `map` onto the state$` observable instead of maintaining multiple lists.
+
+
+```ts
+enum Filter { ALL, ACTIVE, COMPLETED }
+
+interface ITodoStore {
+  filter: Filter;
+  todos: ITodo[];
+}
+
+@Injectable()
+export class TodoStore {
+  private stateSubject$: BehaviorSubject<ITodoStore> = new BehaviorSubject({
+    filter: Filter.ALL,
+    todos: [],
+  });
+
+  public readonly state$: Observable<ITodoStore> = this.stateSubject$.asObservable();
+
+  constructor(private apiService: ApiService) {
+    this.initialize();
+  }
+
+  initialize() {
+    this.apiService.get('/todos')
+      .subscribe(users => {
+        this.stateSubject$.next(
+          Object.assign({}, this.stateSubject$.getValue(), { users })
+        );
+      });
+  }
+
+  filterTodoByType(todo, filterType) {
+    if (filterType === Filter.ALL) {
+      return true;
+    } else if (filterType === Filter.COMPLETED) {
+      return todo.completed;
+    } else {
+      return !todo.completed;
+    }
+  }
+
+  filteredTodos$() {
+    return this.state$.map(({ filter, todos }) =>
+      todos.filter(todo => this.filterTodoByType(todo, filter));
+  }
+}
+```
+
+Observables also allow us to merge multiple streams to implement complex workflows such as a [debounced search which ignores stale requests](../http/search_with_flatmap.md) in a declarative manner.
+
+
+### Subscribing to Data
+To access a data property use operators such as `map` or `pluck` to derive separate streams. This allows you to utilize the `async` pipe in the templates and removes the need for managing subscriptions in the component class.
```

Comment:
Maybe it's worth mentioning that like Promises, observables beget observables, and that you can "cache" a stream for reuse. Any operators or transformations that occurred in that base observable get passed on to any subsequent "child" or "branch" observables